File: 875. Koko Eating Bananas: O(max(p)*n).py
# 875. Koko Eating Bananas
# https://leetcode.com/problems/koko-eating-bananas/description/

# O(max(p) * n) algorithm for finding minimum k bananas per hour to eat n piles of bananas (exceeds time limit):
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        k_minimum = max(piles)
        # Iterate from down from largest k to check if can eat all piles in h hours:
        for k in range(max(piles), 0, -1):
            total_time = 0
            # Each pile takes ⌈p / k⌉ hours (i.e., rounded up):
            for i in range(len(piles)):
                total_time += math.ceil(piles[i] / k)
            # If total eating time fits within hours, update new minimum k:
            if total_time <= h:
                k_minimum = min(k_minimum, k)
        
        # Return found minimum k bananas per hour to eat all piles in h hours:
        return k_minimum

File: 875. Koko Eating Bananas: O(max(p)*n).rb
# 875. Koko Eating Bananas
# https://leetcode.com/problems/koko-eating-bananas/description/

# @param {Integer[]} piles
# @param {Integer} h
# @return {Integer}

# O(max(p) * n) algorithm for finding minimum k bananas per hour to eat n piles of bananas (exceeds memory limit:
def min_eating_speed(piles, h)
    k_minimum = piles.max
    # Iterate from down from largest k to check if can eat all piles in h hours:
    for k in (1..piles.max).to_a.reverse
        total_time = 0
        # Each pile takes ⌈p / k⌉ hours (i.e., rounded up):
        piles.each do |pile|
            total_time += (pile.to_f / k).ceil
        end
        # If total eating time fits within hours, update new minimum k:
        k_minimum = [k_minimum, k].min if total_time <= h
    end
    
    # Return found minimum k bananas per hour to eat all piles in h hours:
    k_minimum
end
